File: adapters/elixir/i18nt.ex
defmodule I18nt do
  @doc """
  获取翻译文本并进行插值
  """
  def t(data, path, params \\ %{}) do
    keys = String.split(path, ".")
    val = get_in(data, keys)

    case val do
      nil -> "[#{path}]"
      str when is_binary(str) ->
        Enum.reduce(params, str, fn {k, v}, acc ->
          String.replace(acc, "{#{k}}", to_string(v))
        end)
      _ -> val
    end
  end
end

File: adapters/scala/I18nt.scala
package i18n

import scala.language.dynamics

/**
 * i18nt Scala Helper Class using Dynamics
 */
class I18nt(val data: Any) extends Dynamic {

  def selectDynamic(name: String): I18nt = {
    data match {
      case m: Map[String, _] @unchecked => 
        new I18nt(m.getOrElse(name, s"[$name]"))
      case _ => 
        new I18nt(s"[$name]")
    }
  }

  def apply(params: Map[String, Any] = Map.empty): String = {
    data match {
      case s: String =>
        params.foldLeft(s) { case (acc, (k, v)) =>
          acc.replace(s"{$k}", v.toString)
        }
      case _ => data.toString
    }
  }

  override def toString: String = data.toString
}
